Add draft post "Operator Constraints"

Merovius opened the pull request:
A rendered version can be viewed here: https://website-review.merovius.de/pull/constraints/posts/draft-operator-constraints/

I'm mainly looking for general understandability and structural feedback, but any comment is welcome.

Merovius committed:
Add draft post "Operator Constraints"

Merovius committed:
Light editing

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
````diff
@@ -0,0 +1,349 @@
+
+```go
+type Integer interface {
+  ~int | ~int8 | ~int16 | ~int32 | ~int64 | ~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr
````
The syntax highlighting on your page makes this look pretty ugly 😬.

Merovius replied:
[Yeah, I know](https://twitter.com/TheMerovius/status/1526285257942638594). I should use a custom syntax highlighting color scheme that is less garish.

Merovius replied:
Changed the syntax highlighting to be differently ugly.

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
```diff
@@ -0,0 +1,349 @@
+3. Taking a comparison function
+4. Comparator types
+
+I believe the fourth is a novel idea. At least I don't think I have seen it
+come up in a discussion about Go so far. Though I think I have seen it used in
+other languages.
```
This paragraph feels a bit out of place here. Maybe move it to the comparator types section?

Merovius replied:
The rationale of that sentence is to entice people to continue reading til the end - that is, I don't want them to open the post and think "okay, same old" and close it again. At the same time, that section must be at the end, so it is clear what it's trying to "fix".

That is, if this reads strange, I'd just remove it. But I'd like some way to make sure the reader gets to that section. Any ideas?

johanbrandhorst replied:
I thought about this again after reading the post and I changed my mind - it is nice to whet the readers appetite.

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
```diff
@@ -0,0 +1,349 @@
+
+## `constraints.Ordered`
+
+Go 1.18 has a mechanism to constraint a type parameter to all types which have
```
```suggestion
Go 1.18 has a mechanism to constrain a type parameter to all types which have
```

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
```diff
@@ -0,0 +1,349 @@
+They can't use your code with predeclared types like `int` or `string` but
+always have to wrap it into a new type.
+
+Likewise, if a type already has a natural comparison method but it is not
```
This comma feels out of place when the sentence is so short. Either remove the comma or merge the first and second sentence.
```suggestion
Likewise if a type already has a natural comparison method but it is not
```

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
```diff
@@ -0,0 +1,349 @@
+useful zero values. Your `SearchTree` type needs the `Less` field to be
+populated to work. So its zero value can not be used to represent an empty set.
+
+You can not even lazily initialize it (which is a common trick to make types
```
```suggestion
You cannot even lazily initialize it (which is a common trick to make types
```
Or `can't`

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
```diff
@@ -0,0 +1,349 @@
+  return a < b
+}
+
+type LessMethod[T Lesser[T]] struct{}
```
Would it be worth defining the `Lesser` interface too?

Merovius replied:
I was trying to implicitly refer back to the one used in the second section. Maybe that's not clear, but I think I'm fine with that.

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
````diff
@@ -0,0 +1,349 @@
+  var c C
+  return c.Less(b, a)
+}
+```
````
Is it possible to define a helper that takes a less function literal?

Merovius replied:
Unfortunately no. Its zero value would not be callable. I will add a sentence to be explicit.

johanbrandhorst commented on content/posts/draft-operator-constraints.markdown:
```diff
@@ -0,0 +1,349 @@
+
+|| `constraints.Ordered` | `Lesser[T]` | `func(T,T) bool` | `Comparator[T]` |
+|--------------------------|----|--------|----|----|
+| Predeclared types        | 👍 | 👎     | 👎 | 👎 |
```
The thumbs up and down are hard to differentiate on your page, maybe use ✅ and ❌?

Merovius replied:
The checkmark and cross read as "yes or no" to me, while the thumbs read as "good or bad". That is, using the thumbs saves the extra step of having to decode which is "better". For example, for "Predeclared types" a "yes" is better, while for "type boilerplate" a "no" is better.

Alternatively, I would have to find phrasings for every line which mean "yes is always better", which also have to be short enough. I failed at that, myself :)

johanbrandhorst reviewed (commented):
This was super interesting! I too agree that this is a novel solution, but also agree that it's a bit rough to use.

Merovius committed:
Apply suggestions from code review

Co-authored-by: Johan Brandhorst-Satzkorn <johan.brandhorst@gmail.com>

Merovius committed:
Add explicit sentence about Comparator[T] and func values

Merovius committed:
Remove claim that comparator types are novel

Just stumbled over the fact that Roger actually suggested them a while
ago:
https://github.com/golang/go/issues/46477#issuecomment-859405752

As the sentence seems to read strangely anyway, let's just remove it.

mdlayher reviewed (approved):
Looks great! The Comparator approach is pretty neat.

Merovius commented:
Published now, closing this.
